ibis/backends/clickhouse/compiler/core.py

```diff
@@ -68,31 +91,52 @@ def translate(op: ops.TableNode, params: Mapping[ir.Value, Any]) -> sg.exp.Expre
     sqlglot.expressions.Expression
         A sqlglot expression
     """
-    params = {param.op(): value for param, value in params.items()}
 
     alias_index = 0
     aliases = {}
 
-    def fn(node, cache, params=params, **kwargs):
+    def fn(node, cache, **kwargs):
         nonlocal alias_index
 
+        result = _translate_node(node, aliases=aliases, cache=cache, **kwargs)
+
+        if not isinstance(node, ops.TableNode):
+            return result
+
         # don't alias the root node
         if node is not op:
             aliases[node] = f"t{alias_index:d}"
             alias_index += 1
 
-        raw_rel = translate_rel(
-            node, aliases=aliases, params=params, cache=cache, **kwargs
-        )
-
         if alias := aliases.get(node):
             try:
-                return raw_rel.subquery(alias)
+                return result.subquery(alias=alias)
             except AttributeError:
-                return sg.alias(raw_rel, alias)
+                return sg.alias(result, alias=alias)
         else:
-            return raw_rel
-
-    results = op.map(fn, filter=ops.TableNode)
+            return result
+
+    # substitute parameters immediately to avoid having to define a
+    # ScalarParameter translation rule
+    #
+    # this lets us avoid threading `params` through every `translate_val` call
+    # only to be used in the one place it would be needed: the ScalarParameter
+    # `translate_val` rule
+    params = {param.op(): value for param, value in params.items()}
+    op = op.replace(
+        p.ScalarParameter
+        >> (lambda op, _: ops.Literal(value=params[op], dtype=op.dtype))
+    )
+
+    # rewrite cumulative functions to window functions, so that we don't have
+    # to think about handling them in the compiler, we need only compile window
+    # functions
+    op = op.replace(
+        p.WindowFunction(p.Cumulative, ...)
+        >> (lambda op, _: cumulative_to_window(op.func, op.frame))
+    )
```

Comment:
The pattern system is pretty darn sweet.

Reply:
Let's avoid building that until we have some more use cases as input

Reply:
Single-pass replacement worked 🎉

Reply:
Note that this will work until the rules have distinct matchers because the first matching pattern gets applied.